cmds/core/cat/cat.go

```diff
@@ -41,6 +41,13 @@ func run(args []string, stdin io.Reader, stdout io.Writer) error {
 		}
 	}
 	for _, file := range args {
+		if file == "-" {
```

Comment:
we're starting to scatter these "see if it is '-'" all over our commands. It's a shame, it is a standard from the 70s (yes, the 70s!) that predates /dev/fd. 

I am wondering -- rather than doing what we see here, might we have a function:
util.StdinFileName(s string)
if s is "-", return the appropriate /dev/fd/0 (or /dev/stdin or whatever) to be opened? Or something like it? Otherwise we're going to see lots of tools handling - differently and that's a shame.

Reply:
I think the testing part became much harder. Currently we are using go interfaces to test stdin/stdout with bytes.Buffer's and it's nice and clean. But if we use something like this
```go
func StdinFileName(s string) string {
    if s == "-" { return "/dev/stdin" }
    return s
}
```

We kinda forced to mess with global state and change StdinFileName function during tests or not test at all.
This can be improved by creating interface:
```go
type Stdin interface {
    Path(s string) string
}
```
that will return prepared file in tmpDir and use it in tests. But I'm not quite sure that overall picture will be better that just check if filename == "-" and use stdin that we already provided to our utility.

I will give it a try on weekend, and create an additional commit to see what I got.

Reply:
I wrote a draft implementation, but it's quite a bit of test to rewrite so I just post important part here:
```go
type stdinFinder struct{}

func (f stdinFinder) path(s string) string {
	if s == "-" {
		return "/dev/stdin"
	}
	return s
}

type finder interface {
	path(s string) string
}

// testing part


type testFinder struct {
	stdinPath string
}

func newTestFinder(t *testing.T) testFinder {
        t.Helper()
	tmp := t.TempDir()
	f, err := os.CreateTemp(tmp, "stdin")
	if err != nil {
		t.Fatal(err)
	}
	defer f.Close()
	return testFinder{stdinPath: f.Name()}
}

func (f testFinder) path(s string) string {
	if s == "-" {
		return f.stdinPath
	}
	return s
}
```

So in tests if we need to use stdin we can write something like this:
```go
	var stdout bytes.Buffer
	testFinder := newTestFinder(t)
	err = os.WriteFile(testFinder.stdinPath, []byte("input to stdin"), 0o666)
	if err != nil {
		t.Fatal(err)
	}
```

So it's a little more tedious that using bytes.Buffer as stdin and stdout. Main function will look something like this if we add command struct.

```go
type command struct {
	stdout io.Writer
	finder finder
	args   []string
}

func (c *command) run() error {
	if len(c.args) == 0 {
		c.args = append(c.args, "-")
	}
	for _, file := range c.args {
		file = c.finder.path(file)
		f, err := os.Open(file)
		if err != nil {
			return err
		}
		if err := cat(f, c.stdout); err != nil {
			return fmt.Errorf("failed to concatenate file %s to given writer", f.Name())
		}
		f.Close()
	}
	return nil

}
```

So anyway we need to handle an empty args case, or use have a reference to os.Stdin. So imho it's not worth it, because we write much more test code than utility code. And gnu cat just did what I did in current comment `bool reading_stdin = STREQ (infile, "-");`

@rminnich  what do you think?

Reply:
Ugh - some things `wget` also use `-` to denote `stdout` (`wget www.google.com -O -`)

Instead of a function that substitutes the filename, maybe we can provide the substitutions of the common os.* function, which return the `*File` structs (which satisfy the `io.Reader` interface) and some other useful functions

```go
package urootfile

import os

func swapName(name) string {
  if name == "-" {
    return "/dev/stdin"
  }
  return name
}

// ReadFile is a substitute for os.ReadFile.
func ReadFile(name string) ([]byte, error) {
  return os.ReadFile(swapName(name))
}

// Open is a substitute for os.Open.
func Open(name string) (*File, error) {
  return os.Open(swapName(name))
}

// OpenFile is a substitute for os.OpenFile.
func OpenFile(name string) (*File, error) {
  return os.OpenFile(swapName(name))
}

```

Then on line 44 of this file you just substitute `f, err := os.Open(file)` for `f, err := urootfile.Open(file)`

Reply:
I think both ways have problem with our current testing approach, the run function have io.Writer that can easily be switched to os.Stdin in main and bytes.Buffer in tests. But if we try to substitute "-" and return os.File or string we need provide an interface that will return os.File or string that are suitable for testing.